src/plot_api/plot_api.js

```diff
@@ -2286,6 +2286,19 @@ Plotly.relayout = function relayout(gd, astr, val) {
             Images.supplyLayoutDefaults(gd.layout, gd._fullLayout);
             Images.draw(gd);
         }
+        else if(p.parts[0] === 'mapbox' && p.parts[1] === 'layers') {
+            Lib.extendDeepAll(gd.layout, Lib.objectFromPath(ai, vi));
+
+            // append empty container to mapbox.layers
+            // so that relinkPrivateKeys does not complain
+
+            var fullLayers = (gd._fullLayout.mapbox || {}).layers || [];
+            var diff = (p.parts[2] + 1) - fullLayers.length;
+
+            for(i = 0; i < diff; i++) fullLayers.push({});
+
+            doplot = true;
```

Comment:
merging this logic with `annotations`, `shapes` and `images` shouldn't be too bad.